File: algorithms/[875]爱吃香蕉的珂珂/solution.ts
// topics = ["二分查找"]

/**
 * Binary Search
 * time O(nlogw), space O(1), n 为香蕉堆的数量, w 为最大堆香蕉的数量
 */
function minEatingSpeed(piles: number[], h: number): number {
  let i = 1;
  let j = Math.max(...piles);

  while (i < j) {
    const mid = Math.floor((i + j) / 2);
    const hour = piles.reduce((acc, cur) => acc + Math.ceil(cur / mid), 0);

    if (hour > h) {
      i = mid + 1;
    } else {
      j = mid;
    }
  }

  return i;
}

File: algorithms/[875]爱吃香蕉的珂珂/solution.py
# topics = ["二分查找"]

from math import ceil
from functools import reduce
from typing import List


class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        i, j = 1, max(piles)

        while i < j:
            mid: int = (i + j) // 2
            hour = reduce(lambda acc, cur: acc + ceil(cur / mid), piles, 0)
            if hour > h:
                i = mid + 1
            else:
                j = mid

        return i
